protocols/bizcoin/bizcoin.go

```diff
@@ -0,0 +1,819 @@
+	bz.tempCommitResponse = append(bz.tempCommitResponse, bzr.Response)
+
+	if len(bz.tempCommitResponse) < len(bz.Children()) {
+		bz.tcrMut.Unlock()
+		return nil
+	}
+
+	if bz.signRefusal {
+		bzr.Exceptions = append(bzr.Exceptions, cosi.Exception{bz.Public(), bz.commit.GetCommitment()})
+		bz.tcrMut.Unlock()
+	} else {
+		resp, err := bz.commit.Response(bz.tempCommitResponse)
+		bz.tcrMut.Unlock()
+		if err != nil {
+			return err
+		}
+		bzr.Response = resp
+	}
+
+	// notify we have finished to participate in this signature
+	bz.doneSigning <- true
+	dbg.Lvl3(bz.Name(), "BizCoin handle Response COMMIT (refusal=", bz.signRefusal, ")")
+	// if root we have finished
+	if bz.IsRoot() {
+		sig := bz.Signature()
+		if bz.onResponseCommitDone != nil {
+			bz.onResponseCommitDone()
+		}
+		if bz.onSignatureDone != nil {
+			bz.onSignatureDone(sig)
+		}
+		bz.Done()
+		return nil
+	}
+
+	// otherwise , send the response up
+	err := bz.SendTo(bz.Parent(), bzr)
+	bz.Done()
+	return err
+}
+
+func (bz *BizCoin) handleResponsePrepare(bzr *BizCoinResponse) error {
+	// check if we have enough
+	bz.tprMut.Lock()
+	bz.tempPrepareResponse = append(bz.tempPrepareResponse, bzr.Response)
+	if len(bz.tempPrepareResponse) < len(bz.Children()) {
+		bz.tprMut.Unlock()
+		return nil
+	}
+
+	// wait for verification
+	bzrReturn, ok := bz.waitResponseVerification()
+	if ok {
+		// append response
+		resp, err := bz.prepare.Response(bz.tempPrepareResponse)
+		bz.tprMut.Unlock()
+		if err != nil {
+			return err
+		}
+		bzrReturn.Response = resp
+	} else {
+		bz.tprMut.Unlock()
+	}
+
+	dbg.Lvl3("BizCoin Handle Response PREPARE")
+	// if I'm root, we are finished, let's notify the "commit" round
+	if bz.IsRoot() {
+		// notify listeners (simulation) we finished
+		if bz.onResponsePrepareDone != nil {
+			bz.onResponsePrepareDone()
+		}
+		bz.startChallengeCommit()
+
+		return nil
+	}
+	// send up
+	return bz.SendTo(bz.Parent(), bzrReturn)
+}
+
+// computePrepareResponse wait the end of the verification and returns the
+// BizCoinResponse along with the flag:
+// true => no exception, the verification is correct
+// false => exception, the verification is NOT correct
+func (bz *BizCoin) waitResponseVerification() (*BizCoinResponse, bool) {
+	bzr := &BizCoinResponse{
+		TYPE: ROUND_PREPARE,
+	}
+	// wait the verification
+	verified := <-bz.verifyBlockChan
+	if !verified {
+		// append our exception
+		bzr.Exceptions = append(bzr.Exceptions, cosi.Exception{bz.Public(), bz.prepare.GetCommitment()})
+		bz.sendAndMeasureViewchange()
+		return bzr, false
+	}
+
+	return bzr, true
+}
+
+// verifyBlock is a simulation of a real verification block algorithm
+func verifyBlock(block *blockchain.TrBlock, lastBlock, lastKeyBlock string, done chan bool) {
+	//We measure the average block verification delays is 174ms for an average
+	//block of 500kB.
+	//To simulate the verification cost of bigger blocks we multiply 174ms
+	//times the size/500*1024
+	b, _ := json.Marshal(block)
```

Comment:
I think one would have to keep tracks of transactions already validated ? and verify if the hash has a leading numbers of '0' in front ? Seems more complicated than what's needed for the moment, but could be done, sure.
I don't know about the paper.